### PR TITLE
8308235: ThreadContainer registry accumulates weak refs

### PR DESCRIPTION
**Pull request by AlanBateman**
ThreadContainers is an internal class used to make thread pools and other groupings of threads discoverable for observability. Some refactoring in 2021 (in the loom repo, and before integration) accidentally changed the creation of a weak reference so that it no longer associated with the reference queue. The result is that stale refs aren't expunged from a CHM, leading to a memory leak. The change to fix the issue is trivial.

Tests for memory leaks can be problematic, often more trouble than they are worth. I started with a test that polls the size of the internal CHM but decided to ditch it. Instead, the test is simple. It just runs for a few seconds creating ExecuorService implementations (including TPE, TPPE, and FJP), unreferencing them without shutdown (so they don't terminate and unregister). This is enough to causes OOME with product builds a small heap.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308235](https://bugs.openjdk.org/browse/JDK-8308235): ThreadContainer registry accumulates weak refs

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [89475604](https://git.openjdk.org/jdk/pull/14047/files/89475604c6aa97880c57839e9d4ce9f6610ec4d3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14047/head:pull/14047` \
`$ git checkout pull/14047`

Update a local copy of the PR: \
`$ git checkout pull/14047` \
`$ git pull https://git.openjdk.org/jdk.git pull/14047/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14047`

View PR using the GUI difftool: \
`$ git pr show -t 14047`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14047.diff">https://git.openjdk.org/jdk/pull/14047.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14047#issuecomment-1554214633)

### CODE REVIEW

File: test/jdk/java/util/concurrent/Executors/UnreferencedExecutor.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * @run main/othervm -Xmx32m UnreferencedExecutor
+ */
+
+import java.util.concurrent.Executors;
+
+public class UnreferencedExecutor {
+    public static void main(String[] args) throws Exception {
+        int ncores = Runtime.getRuntime().availableProcessors();
+        long start = System.currentTimeMillis();
+        while (System.currentTimeMillis() - start < 5000) {
```
Comment (viktorklang-ora) [inline test/jdk/java/util/concurrent/Executors/UnreferencedExecutor.java:38]: @AlanBateman Might want to have this based on `nanoTime()` such that changes to wall-clock doesn't impact the length of the run of the test.

**Comment by jaikiran** [comment]
The change looks OK to me. The `QUEUE` (of `WeakReference`) and thus the registry, gets cleaned up whenever new `ThreadContainer`s are registered with the registry.

Would it be worth calling `expungeStaleEntries()` to do this cleanup even when a specific `ThreadContainer` is deregistered, in the implementation of method `deregisterContainer(...)`?

**Review by jaikiran — APPROVED** [review]

File: test/jdk/java/util/concurrent/Executors/UnreferencedExecutor.java (hunk 1 above)
Comment (AlanBateman) [reply test/jdk/java/util/concurrent/Executors/UnreferencedExecutor.java]: Okay, that would be better.

**Pull request closed without merging**